crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,341 @@
+//! Fetch is a Kitsune2 module for fetching ops from peers.
+//!
+//! In particular it tracks which ops need to be fetched from which agents,
+//! sends fetch requests and processes incoming responses to these requests.
+//!
+//! It consists of multiple parts:
+//! - Data object that tracks op and agent ids in memory
+//! - Fetch task that requests tracked ops from agents
+//! - Incoming op task that processes incoming responses to op requests by
+//!     - persisting ops to the data store
+//!     - removing op ids from in-memory data object
+//!
+//! ### Data object [`CoreFetch`]
+//!
+//! - Exposes public method [`CoreFetch::add_ops`] that takes a list of ops and an agent id.
+//! - Stores pairs of ([`OpId`][`AgentId`]) in the order that ops have been added.
+//!
+//! ### Fetch task
+//!
+//! - Toeks pairs of ([`OpId`], [`AgentId`]) from the beginning of the queue and requests the op from the agent.
+//! - Appends removed pair to the end of the queue.
+//! - Put unresponsive peers on a cool-down list.
+//!
+//! ### Incoming op task
+//!
+//! - Incoming op is written to the data store.
+//! - Once persisted successfully, op is removed from the data object.
+
+use std::{
+    collections::HashMap,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use indexmap::IndexMap;
+use kitsune2_api::{
+    builder,
+    config::ModConfig,
+    fetch::{DynFetch, DynFetchFactory, Fetch, FetchFactory},
+    tx::Transport,
+    AgentId, BoxFut, K2Result, OpId,
+};
+use tokio::{
+    select,
+    sync::{mpsc::Sender, Mutex},
+};
+
+const MOD_NAME: &str = "Fetch";
+
+/// Configuration parameters for [CoreFetchFactory].
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CoreFetchConfig {
+    /// How many parallel op fetch requests can be made at once. Default: 2.  
+    parallel_request_count: u8,
+    /// Duration in ms to pause when parallel request count is reached. Default: 10.
+    parallel_request_pause: u64,
```

Comment:
It would be nice if this worked with blocking and waiting for capacity to be available. I'm gussing at this point but I'm assuming I'm going to see a sleep with this value further down